### PR TITLE
Bookmarks: view and edit

### PR DESCRIPTION
**Pull request by hius07**
Annotations are: bookmarks, highlights, notes.

Changes proposed:

(1) Annotations list: show page number and timemark.
(2) Rename annotaion: show page number and timemark in the input dialog description.
Page numbers in (1) and (2) reflects text layout changes (font size etc).

(3) Having (1) and (2), there is no need to add page number and timemark to the annotation `text` field. Closes https://github.com/koreader/koreader/issues/8211.
(4) Updating highlight boundaries doesn't kill the annotation `text` field. Closes https://github.com/koreader/koreader/issues/6458.
(5) Add bookmark: put the standard 'star' to the annotation `text` field, to distinguish bookmarks from other annotations.

<kbd></kbd>

<kbd></kbd>

<kbd></kbd>

### CODE REVIEW

**Comment by poire-z** [comment]
No strong personnal opinion on this new look vs the current look.
I just feel some people may not much care about the page number and date - and having the page number and date take a good amount of the available width leaves only half the screen for the text, which gets smaller and harder to read for people who care only about it. Having them appended at the end of the text made them share the full available width, which feels like a better all-around solution).

> (5) Add bookmark: put the standard 'star' to the annotation text field, to distinguish bookmarks from other annotations.

Didn't we use to show the chapter the bookmarked page is in ? Feels like more useful information than just a star.

----
Some more technical notes:
It looks like you're mostly just tweaking bookmarks copies for display, so not hacking saved bookmarks, in the first part of the PR.
But in updateBookmark(), it seems you do (but I'm not sure): no more saving the (ugly) `page.text`, right? And adding a new `page_str` ?

But we have the ambitious and forgotten #5794 that should make our bookmarks/annotations better.
I fear doing any kind of tweaks to the current bookmarks structures will make this big cleanup harder, as we will have to detect bookmarks before this PR and after this PR to do a different migrations.
Do you think your changes will cause some complications?

Or do you think/feel we can just go on with small tweaks and forget about the ambitious refactor (that no one seems to want to take over...)

Also, please test your changes with the book in the first post of #5766 : it's a document with hundreds of highlights, See that turning pages and showing the bookmark list doesn't get slower before/after your changes.
And that it works as well with PDF.

**Comment by poire-z** [comment]
> But in updateBookmark(), it seems you do (but I'm not sure): no more saving the (ugly) page.text, right? And adding a new page_str ?

You also need to think if it may cause disruptions with the few other softwares that may parse KOReader's highlights, like @noembryo 's KOHighlights , and expect `text` and don't know about `page_str`

**Comment by hius07** [comment]
Thank you for the questions. In order of appearance (not of importance):


> some people may not much care about the page number and date

And some do care. Showing page / timemark can be done switchable in settings.


> Having them appended at the end of the text

Page number should be actual after changing text layout, as asked in https://github.com/koreader/koreader/issues/8211.


> Didn't we use to show the chapter the bookmarked page is in ? Feels like more useful information than just a star.

Chapter is added (and is still added) to the `notes` field. It is not seen in the list of bookmarks.


> tweaking bookmarks copies for display

Yes, having a deep copy to show the list, to build `item_table` with mandatory.


> in updateBookmark(), it seems you do (but I'm not sure): no more saving the (ugly) `page.text`, right?

Yes, changing highlight boundaries doesn't change the `text` field.


> And adding a new `page_str` ?

No, `page_str` is added to the `item_table` only. It is not added to bookmarks and is not saved in metadata.


> But we have the ambitious and forgotten #5794 that should make our bookmarks/annotations better.

There are so many open FR about bookmarks! People are suffering waiting for the global changes. Let's help them a little bit.


> please test your changes with the book in the first post of #5766

Done, 20 pages of bookmarks, works fine.

> may cause disruptions with the few other softwares that may parse KOReader's highlights

No changes to the bookmarks format in metadata.

> and expect `text` and don't know about `page_str`

As previously, the `text` field is filled with `notes` if empty. No `page_str` exists in the bookmarks.

To summarize:
-technically: the proposed changes doesn't touch the logic and format. There are changes in showing the list and rename dialog only.
-generally: while waiting for the great revision https://github.com/koreader/koreader/issues/5794, we can make some requested enhancements, like actual page numbers or multiple removing of bookmarks, or sorting.

**Comment by poire-z** [comment]
> > Having them appended at the end of the text
> Page number should be actual after changing text layout, as asked in #8211.

You could still append the "actual" page number to the text value, at display time, just concatenating them all in a single full-width text, instead of making them into 2 parts (text | mandatory).
But for past bookmarks/highlights, you'll still show `Page 12 the real text @2021-10-05...` on the left, and `12 2021-10-05...` on the right, no? (You're not cleaning past highlights text, which will be one of the tough task of the annotations refactoring..)

> Chapter is added (and is still added) to the notes field. It is not seen in the list of bookmarks.

Well, I do see a chapter title in the bookmark list, when I just bookmark a page.

> No, page_str is added to the item_table only. It is not added to bookmarks and is not saved in metadata.

Ok, I missed the `bookmark = require("util").tableDeepCopy(bm)`. Having it named `bookmark` doesn't help noticing it's just a copy, unlike `item_table` :)

> Yes, changing highlight boundaries doesn't change the text field.

Even when I have never modified the initial text by adding some personal notes ? If I extend my highlight boundaries, will the initial text not be updated? If that's right, I think it's a defect. (and I know it's hard to guess if the current text is a automatic text, or some previously hand edited text...)
It's really hard to get into this code - ReaderBookmark may be called by ReaderHighlight, they have copies of the same stuff, we update text/notes and it's not really obvious what's in which...
So "notes" is the real book text - and text is just something we make out for display purpose (and save it in metadata which wasn't really needed) - and it happens to also hold the user edited text (which make it needed to be saved in metadata).
I'm going to trust you on this, because I don't want to put my brain on re-understanding the whole code :)

> No changes to the bookmarks format in metadata.

Well, good, that's really all I care about - except for the previous point.

> There are so many open FR about bookmarks! People are suffering waiting for the global changes. Let's help them a little bit.

Well, I was hoping this accumulation and suffering would at some point force someone to take over the huge refactoring.
Thanks for this chocolate box and the thoughts, but that's not a cure. And with too much chocolate, we may forget we have a sickness to cure :)

**Comment by hius07** [comment]
> You're not cleaning past highlights text

No cleaning, no migration.
Well, I really do not like these addings of Page ... to the text, but maybe there are fans of them, so my propositions should be dropped.



> I do see a chapter title in the bookmark list, when I just bookmark a page.

Oh, yes, the `text` is filled with `notes` if empty. Anyway, some mark to distinguish bookmarks might be good?
'Star' + chapter name?


> If I extend my highlight boundaries, will the initial text not be updated? If that's right, I think it's a defect.

An opposite opinion: https://github.com/koreader/koreader/pull/8301
After proposed changes, `text` is kept, and `notes` (visible in Bookmark details) is changed to show actual highlight.

> force someone to take over the huge refactoring

No progress for almost 2 years.

> but that's not a cure

So, dropping?

**Comment by NiLuJe** [comment]
Not a feature I use extensively (or, at all, to be frank), so, hard to comment on any of it, really :/.

**Comment by hius07** [comment]
> the timestamps look weird and large

A compromised solution might be to show only a page number in the list and to show both page number / timestamp in the rename dialog.

The main question remains: is it good to add the page number / timestamp into the text of the bookmark automatically.

**Comment by poire-z** [comment]
> Well, I really do not like these addings of Page ... to the text, but maybe there are fans of them, so my propositions should be dropped.

I also do not like them. I'd rather just see the plain text as-is, with date + page either appended at the very end (at display time), or in an additional column like you did (if it wouldn't take as much space... 2 rows like we have on the right for Detailed list in File browser would be better, but then, it's much more work, Menu+mandatory is not enough).

But well, we inherited this whole crap...

> Anyway, some mark to distinguish bookmarks might be good?
'Star' + chapter name? (The star is taken from the Skim widget).

Sure, I'm with you on this, easier to distinguish page bookmarks from highlights. A star is fine (but a dogear would be better, if we had a corner to stick it too, which we don't really have here...)

> > If I extend my highlight boundaries, will the initial text not be updated? If that's right, I think it's a defect.
> An opposite opinion: #8301

I guess you mean #6458 as the opposite opinion ?
But I agree with it that we shouldn't trash user-updated text.
But while fixing that, we shouldn't break something else, and not-update automatic-updated text.
There's obviously a conflict there - because we re-use a same crappy field for both...

> After proposed changes, text is kept, and notes (visible in Bookmark details) is changed to show actual highlight.

You mean there is no issue and automatic-updated text is updated ? Hard to follow :)

> So, dropping?

Don't close too fast, we're dicussing, throwing ideas, foreseeing issues and use cases we need to be able to handle.
You have 5 points in your first post
- (2) and (5) are probably good and valid and risk-free
- (1) is aesthetic preference
- (3) feels migration-risky
- (4) feels like it breaks a use case for solving another one 

A thought about (4): may be we can rebuild a "Page 12 real text date" from `notes` and the other fields at display time/before modification, and if it is equal to what's in the `text` field (so, no update, no UI language change), we could consider it "automatic text" and ignore/trash/update it, and build a dynamic one like you did from the other fields ? If there's a difference, consider it "user-updated text".
I guess something like that would have to be done for the big migration.

> No progress for almost 2 years.

Yes, it's sad.
I'm sometimes considering it, but I know I'll need a few weeks of no-distraction - and I'd need to re-read all the linked issues... and I always find some smaller more fun issue to play with :)

But once the datastructure is updated/migrated, and the code that updates them is fine, there could be many ways to show/update bookmarks and highlights. I'm pretty sure we can't find a single one that will please everyone, so we'll probably get plugins that would show/sort them different ways (or else, we'd have a huge Bookmark settings> menu like we have for the footer, to allow for the various ways people want to use/see bookmarks).
Until we have that, everything is limited and frustrating, and some of the cases you want to solve will just feel like half made or break someother use case/habit :/

> A compromised solution might be to show only a page number in the list and to show both page number / timestamp in the rename dialog.

I might be fine with that, the page number would make the mandatory small enough.

> The main question remains: is it good to add the page number / timestamp into the text of the bookmark automatically. (I really hate it but tastes differ).

Obviously it is not and we all hate it :) But we may need to keep it for the time being because:
- other software like KOHighlight may expect it
- we may want to prefer having it stay /be updated like this if it can help the future migration and be used for extracting things or deciding if it's automatic or user-edited text. I have no idea if we'll really need it (so, having no idea, my lazy advice is: keep it until one of us can really give it some thoughts and answer this question :)

**Comment by hius07** [comment]
<kbd></kbd>

**Comment by hius07** [comment]
Changes proposed (updated):

(1) Annotations list: show page number.

(2) Rename annotaion: show page number and timestamp in the input dialog description.
Now allows blank input: resets `text` field to auto-text (see 3).

(3) If empty, `text` field is filled when:
-showing Bookmarks list
-calling Edit highlight dialog
New setting `Add page number / timestamp to bookmark` (enabled by default, keeping status quo).
If enabled, empty `text` field will be filled with auto-text (Page # _notes_ @ datetime).
If disabled, empty `text` field will be filled with just the `notes` field.

(4) Updating highlight boundaries: if the `text` field has been edited manually by a user after highlight creation, it will be kept intact.
If not, the `text` field will be updated according to the highlight new boundaries and (3).
Note: after changing highlight boundaries datetime of the annotation is not changed (in readerhighlight module).

(5) Add bookmark: put the standard 'star' to the first position of the annotation `notes` field.
Then the `text` field will be filled in accordance with (3).

**Comment by hius07** [comment]
Do we need separators (switchable)?

<kbd></kbd>

**Comment by hius07** [comment]
> What's the current ordering ?

By page, descending (opposite to ToC).

File: frontend/apps/reader/modules/readerbookmark.lua [hunk 1]
```diff
@@ -681,17 +670,17 @@ function ReaderBookmark:toggleBookmark(pn_or_xp)
...
         -- build notes from TOC
-        local notes = self.ui.toc:getTocTitleByPage(pn_or_xp)
-        local chapter_name = notes
-        if notes ~= "" then
+        local notes = "★" -- distinguish bookmarks from highlights and notes
+        local chapter_name = self.ui.toc:getTocTitleByPage(pn_or_xp)
+        if chapter_name ~= "" then
             -- @translators In which chapter title (%1) a note is found.
-            notes = T(_("in %1"), notes)
+            notes = notes .. " " .. T(_("in %1"), chapter_name)
```
Comment (poire-z) [inline frontend/apps/reader/modules/readerbookmark.lua:677]: This I don't like - and you broke your contract of not updating the datastructure :/

This star is cosmectic/display stuff - you should not have to update it in the datastructure - because then, it becomes just what "Page N  @ date" is in highlights: crap we will have to parse and clean.
You should keep saving page bookmarks how they are, and find a way to distinguish them at display time, and add the stars into what's being used for display purpose (your `item_table` clone).

(And may be `"notes="` has no purpose for page-bookmarks and may be it won't even be used for migrating - but I don't know, and don't want to be forced to investigate that now :)

**Comment by hius07** [comment]
<kbd></kbd>

<kbd></kbd>

<kbd></kbd>

**Comment by noembryo** [comment]
>depending on some option `bookmark.text` may no longer contain the prefix "Page 123" and suffix "@ date" (the other fields are not changed and available as they were before)

I use the prefix/suffix in a regex to extract the original `bookmark.text` and then to compare it with the `highlight.text`.
If they differ, I know that there is a user comment there.
No problem if we ditch them, but I have to find a way to know that in advance, so I don't bother with the regex. Maybe using the `cre_dom_version`?

>`bookmark.text` could already be empty for you previously, and if was filled with "Page 123 note @ date" only when the user invoke Show bookmarks - it may stay empty with this new option (as it would just be a copy of `bookmark.notes`)

I don't understand. Will it be empty or the same as `bookmark.notes`?
I think it can work as well. If its empty or the same as `bookmark.notes`, then no comment from the user there..

>`bookmark.text` can as before contain some user own notes instead of "Page 123 note @ date"

If its different than the `highlight.text`, then its a comment. Also OK.

>the original selected text from the book is still available as `bookmark.notes`

Which is the same as the `highlight.text` I think, so OK too.

**Comment by poire-z** [comment]
This PR adds an option to dicth Page 123 and @ date for people who don't want to see them in KOReader's bookmark list view. It's an option, the default is as it was before (with "Page 123... @ date").

I was hoping you would need no change on your side, as it's mostly just an interim (months...) solution to make bookmarks nicer for some KOReader's users (see screenshots above, which are indeed nicer).
If it could cause some issue or need a KOHighlight update, we could add a message when the user toggles this option (although I'd rather not have to do this).

> but I have to find a way to know that in advance, so I don't bother with the regex. Maybe using the cre_dom_version?

No, nothing to do with cre_dom_version: we won't have it changed with this PR: this new option would work with older cre_dom_version books, and with PDFs.
But with this PR, the fact a highlight text is not a user comment is either one of the three cases:
- text == "Page123 + notes + @ date"     (already possible before, still possible)
- text == "" (empty)    (already possible before, still possible)
- text == notes          (new possibility)

I don't think users have been themselves removing "Pages 123" and "@ date" to make their highlights cleaner :) but if they did, well, it's not a user comment, just a user cleanup :) so the 3rd rule above is met.

> I don't understand. Will it be empty or the same as bookmark.notes?

"text" could/can be empty.
I mean "text" could already have been empty (actually: non existant): when a highlight is just made, its "text" is empty/non-existant. It is only filled with "Page 123" .. "@ date" when the user do the action in KOReader of showing the bookmark list. If some of your users forgot to do that before syncing with KOhighliths, you could already have gotten empty/non-existant "text" for all these years. So, may be you were already handling this case ? (obviously, no user comment then :)

This PR makes it so that, when the new option is enabled, text="" will stay empty even after KOReader shows the bookmark list.
If you rather have the `text` filled with exactly `notes` instead of being empty, we can do that.

> Which is the same as the highlight.text 

Oh, indeed, I just verified this: `bookmark.notes == highlight.text`, and they keep being equal. So when I write `bookmark.notes` above, you can read `highlight.text` :)

**Comment by noembryo** [comment]
>But with this PR, the fact a highlight text is not a user comment is either one of the three cases:
>- text == "Page123 + notes + @ date" (already possible before, still possible)
>- text == "" (empty) (already possible before, still possible)
>- text == notes (new possibility)

Yes, you're right. I _was_ checking for the "no text" case.
So the only case missing is the third which is easy (already did it), but it _will_ need program update.. :cry:

**Comment by poire-z** [comment]
(I guess we could also have done like your regex trick to strip Page 123 and @ date only at view time, - and keep them in what is saved.)

**Comment by noembryo** [comment]
>And this would need no change on your side - and only a small update to this PR.

This would be nice.

And something about the `bookmark.notes == highlight.text`.
Looking at my code, I see that I have a check for equality between them and I wonder if there was a time that this was not the case..
If it was always equal, then I can remove it.

**Comment by poire-z** [comment]
> I wonder if there was a time that this was not the case..
If it was always equal, then I can remove it.

I have no idea :/ It may have been different in the past, I don't know. I can tell what happens with the current ugly code - but having to look at the ugly history of this ugly code is above my pain level acceptance :)
So, you better keep it.

**Comment by hius07** [comment]
> But with this PR, the fact a highlight text is not a user comment is either one of the three cases:
> 
> * text == "Page123 + notes + @ date"     (already possible before, still possible)

A user has changed font size, then edited 'Page 123' to the new actual page.
`text` field looks like not a user comment, but it is a user comment.

What is the purpose for KOHighlight to know wether a user has changed the `text` field?

**Comment by hius07** [comment]
> Or @hius07 could make sure that with the new option, text stays empty/non existant when it's unchanged and would be equal to notes.
> Which is the already the case (I think), except in renameBookmark - see [#8301 (comment)](https://github.com/koreader/koreader/pull/8301#discussion_r724434289).
> And this would need no change on your side - and only a small update to this PR.

Can you please clarify a little bit.

With a new option unchecked (no auto-text), text field of highlight or page bookmark is nil until bookmark list is shown (or Edit in highlight popup menu is called).
After that text field is equal to notes field.

**Comment by hius07** [comment]
Summary of changes

(1) Bookmark list:
-page numbers are displayed
-page bookmarks are marked with a star

(2) Rename bookmark dialog:
-page number and timestamp are displayed in the inpupt dialog description
-blank input renames bookmark to the default name in accordance with (3)

(3) Bookmark name auto filling (`text` field):
-new setting `Add page number / timestamp to bookmark` (default state: enabled)
If disabled, bookmark name is equal to the `notes` field.
If enabled, bookmark name is 'Page # `notes` @ timestamp'.

(4) Changing boundaries of the highlight: the name of the highlight is not changed if it was previously edited by a user.

---
Note to (3). The logic of filling procedure is not changed:
Highlight: `notes` is equal to the highlighted text.
Page bookmarks: `notes` is empty when there is no chapter name (it is the only case when it is nil),
otherwise it is 'in `chapter-name`'.
Empty `text` field is auto filled in accordance with (3) when showing Bookmark list or calling Edit from the highlight popup menu.

**Comment by hius07** [comment]
> > With a new option unchecked (no auto-text), text field of highlight or page bookmark is nil until bookmark list is shown (or Edit in highlight popup menu is called).
> > After that text field is equal to notes field.
> 
> You're sure ?

The lines 583, 592 are from the `updateBookmark` method that is called when changing highlight boundaries only.

I talked about showing bookmark list
https://github.com/koreader/koreader/blob/56d828615275e704617d19f8a7a62fe240b4cacb/frontend/apps/reader/modules/readerbookmark.lua#L335-L337
or calling Edit from the highlight popup menu
https://github.com/koreader/koreader/blob/56d828615275e704617d19f8a7a62fe240b4cacb/frontend/apps/reader/modules/readerbookmark.lua#L609-L611

**Comment by poire-z** [comment]
OK, you're right.
So, any issue for you with ALWAYS setting text to nil (or removing it, as that is the case KOHightlight is already handling) when auto_text is disabled and text would be ==notes , and just you using `item_table.text = notes`  for display ?

**Comment by hius07** [comment]
> any issue for you with ALWAYS setting text to `""`

When exporting (https://github.com/koreader/koreader/issues/7756), a user expects that he has a text, but will get nil.

Currently the only change to KOHightlight is that it thinks that `text` without Page # is a user comment.

**Comment by hius07** [comment]
> any issue for you with ALWAYS setting text to `""`

Also an unexpected result will be from the Exporter plugin
https://github.com/koreader/koreader/blob/23be2f278c66b6467101cdadd4c2c7881b31ca85/plugins/exporter.koplugin/main.lua#L383-L385

**Comment by hius07** [comment]
> > Also an unexpected result will be from the Exporter plugin
> 
> So, may be it should be fixed, as it's currently already an issue, right?

Currently it is not an issue, a user gets what he sees.
It would become an issue if (as you propose) we would display `notes` but export `text`.

**Comment by hius07** [comment]
> I guess the translation is for when displaying the title but still is saved as `Page123`?

It is translated

https://github.com/koreader/koreader/blob/56d828615275e704617d19f8a7a62fe240b4cacb/frontend/apps/reader/modules/readerbookmark.lua#L840

**Comment by hius07** [comment]
> I guess the translation is for when displaying the title but still is saved as `Page123`?
> I have to check this..

If some changes will be made, maybe add a check: if `text` == `notes`, it is not a user comment.

**Comment by hius07** [comment]
> > I guess the translation is for when displaying the title but still is saved as `Page123`?
> > I have to check this..
> 
> If some changes will be made, maybe add a check: if `text` == `notes`, it is not a user comment.

I mean, the proposed check will fix not the issue with translation, but our issue with a new design of the `text` field (lack of 'Page #').

**Comment by hius07** [comment]
> The exporter plugin will see the bookmarks with text=nil - and if it doesn't go check `notes`, it will export no text. So, isn't this a current bug ?

Agree, it is not a good behaviour of the plugin.
Let's return to the plugin after the great refactoring.

**Comment by noembryo** [comment]
>Is it so bad that the clear text without 'Page #' will be treated as a user comment?

If the `Page #`/`Date` is missing and some text exists, then yes, it will be treated as user comment.

>I do not know the program, as I understand the user comments are just shown in a separate field?

They have a little different formatting, but the main problem is that they can be edited and then saved in the metadata.

**Comment by poire-z** [comment]
@hius07 : I have told you why I would prefer having text=nil.
You haven't tell us the reasons you would prefer the other way. Can you ?
(And if wanting text=notes, you still have the inconsitency I mentionned before: updateBookmark (moving highlight boundaries) will reset text to nil.)

**Comment by hius07** [comment]
> would prefer having text=nil.

Okay, I'll look at it.

**Comment by hius07** [comment]
> getBookmarkAutoText also when the option is off and we know it's a no-op

Trade off to have only one check of `G_reader_settings:nilOrTrue("bookmarks_items_auto_text")` in the code.

I've tested on the heavy book from https://github.com/koreader/koreader/issues/5766, seems to work fine.

**Comment by poire-z** [comment]
OK.
Don't you want to have a quick look at the exporter plugin and have it fallback to using notes when text is nil ? (Haven't checked if it's complicated, so up to your feeling :)

**Comment by hius07** [comment]
> Don't you want to have a quick look at the exporter plugin and have it fallback to using notes when text is nil ?

Sorry for my alarm, it seems that there is no issue there with empty `text` in bookmark.
It takes clipping.text from the highlight `text` (the same text we have in the bookmark `notes`)
https://github.com/koreader/koreader/blob/23be2f278c66b6467101cdadd4c2c7881b31ca85/plugins/exporter.koplugin/clip.lua#L231-L237
and then checks bookmark, if no `text` in bookmark, just keeps the text from the highlight.
https://github.com/koreader/koreader/blob/23be2f278c66b6467101cdadd4c2c7881b31ca85/plugins/exporter.koplugin/clip.lua#L239-L244

**Comment by poire-z** [comment]
I'd rather like we wait after the next release to merge this PR.
@Frenzie : when do you plan it?

**Comment by Frenzie** [comment]
Sometime next week, I haven't had much time to test or prep release notes yet. Also we won't have an illustration until sometime next week. :-)

**Comment by hius07** [comment]
I can add multiple removing (https://github.com/koreader/koreader/issues/8038#issuecomment-938524080).

**Comment by hius07** [comment]
Final summary of changes

(1) Bookmark list:
-page numbers are displayed
-page bookmarks are marked with a star

(2) Rename bookmark dialog:
-page number and timestamp are displayed in the inpupt dialog description
-blank input renames bookmark to the default name in accordance with (3)

(3) Bookmark name auto filling (`text` field):
-new setting `Add page number / timestamp to bookmark` (default state: enabled)
If disabled, bookmark name is equal to the `notes` field.
If enabled, bookmark name is 'Page # `notes` @ timestamp'.

(4) Changing boundaries of the highlight: the name of the highlight is not changed if it was previously edited by a user.

(5) Sorting of bookmarks list by page number descending or ascending.

**Comment by hius07** [comment]
Trying to increase (not only) bookmark font size people with poor eyes cannot increase shrinked items.

@poire-z, how do you feel about changing
https://github.com/koreader/koreader/blob/91b921505c24e9024295cbaf2c428359a2d82dc3/frontend/ui/widget/menu.lua#L364
to something like `min_font_size = self.font_size - 6`?
Or add a setting `Minimal shrinked item font size`.

**Comment by hius07** [comment]
Currently, bookmark font size = 24

<kbd></kbd>

-6 (seems not enough shrinked)

<kbd></kbd>

-8 (seems a good balance, well shrinked, not too small)

<kbd></kbd>

**Comment by poire-z** [comment]
Looks ok with -8.
Can you show with bookmark font size = 20 (so, with -8, we'll get min font size =12 as before) ?
And with bookmark font size = 16 (which is what I have on my Kobo), see that 16-8 = 8 is not too-too small, in which case you might still want to cap min_font_size to 12, or 10 (or 8 if it's ok but would become too small with bookmark font size = 12).

(I thought somewhere we had a hard cap to the min real font size, but I can't find it - may be it's only in crengine that I've seen it.)

**Comment by hius07** [comment]
20

<kbd></kbd>

16

<kbd></kbd>

I think of `math.max(12, self.font_size - 8)`
